### PR TITLE
Discover Rebuild

### PR DESCRIPTION
**Pull request by RyRy79261**
## Tl;Dr:

This overhaul implements Redux-Saga for improved state & business logic management, code generators, theming engine, caching & performance improvements as well as token data syncing for future development.

## Front end changes
The entire webapp from the ground up has been rebuilt, all is easily customisable and future changes should be more straight forward. 

The rebuild was aimed to allow the app to be mobile responsive from the get go.

## Data flow changes

Onchain state has been prioritised, while caching server has had no modifications, additionally, data is pulled from the target network regardless of browser's Web3 capabilities or selected network.

#### Potential improvements
* Namespace the caching server, refactor the stored data/structure on said server. 
* Decouple the onchain fetching & caching server fetching from the data polling

### Contracts
All contract functions are now selfcontained callable functions.

#### Potential improvements
* Typechain replacement for contracts

### Theming

Theming is currently controlled via Material-Ui, `theme.ts` contains all the theming constants & colors used in the app

#### Potential improvements
* Material replaced with a custom theming configuration & custom component library

### Performance 
* The redux store is now cached to localstorage, allowing for quicker revisit loading times. 
* Reselect has been used to memoize the state selectors

#### Potential improvements
* Treeshaking/selective caching of redux store

### DX (Developer Experience)

A number of changes have been made to improve the overall experience of updating & developing in Discover: 
* A simple monorepo structure has been implemented to allow for single command initialisation of the app
* Component generators via Handlebars.js & Plop allow for creation of `component` `container` & `domain` creation to reduce human error and allow for enforcing of code style preferences via `cd ./WebApp && yarn generate`
* Improved typing, folder structure, constants management has been implemented, `theme.ts`, `utils/`

#### Potential improvements
* Overhauling the monorepo in to a package structure using Rollup
* Refactor the linting settings
* Move most of the functionality in HOC based containers to their child components using Hooks
* Update/Expand templates
* Precommit prettifier via `prettier-quick` 

### Environment variables
Some modification to the environment variables have been made, `.env.example` can be found in each of the sections `Backend`, `Contracts` & `WebApp`

### Issues with existing Embark configuration
I experienced odd behavior from Embark in deploying contracts to ropsten & setting the addresses in the output Embark contract files, as such, work arounds have been implemented while using the output from the existing Embark section of the code base

### Build considerations 
There is an issue (potentially only present on my machine) where `wsrun` doesn't carry over the `NODE_ENV` env variable correctly, thus to build, please update your `.env` settings as needed, then run `yarn build` 

The build in `./WebApp` outputs a zipfile containing the built webapp files, which are ready to run from the root of a serving http server

I am unsure of the requirements or usage of some of the commands within the Contracts & Backend section, changes might need to be made based on feedback

### CODE REVIEW

**Comment by jakubgs** [comment]
There's so many changes here we might as well start a new branch using `git checkout --orphan` or even a new repo.

**Comment by jakubgs** [comment]
Seriously, this looks like a rewrite more than an update. A new repo would make more sense.

**Comment by jakubgs** [comment]
I guess the contract stuff stays the same, but still. I wish there was "mark all renamed as viewed" or something.

File: package.json [hunk 1]
```diff
@@ -1,53 +1,107 @@
...
+    "fullbuild:webapp": "mkdir full-build/frontend && cp -r WebApp/build/* full-build/frontend/",
+    "fullbuild:archive": "cd full-build/ && bestzip ../app.zip * && cd ..",
+    "create:fullbuild": "yarn build && yarn fullbuild:clean && yarn fullbuild:init && yarn fullbuild:misc && yarn fullbuild:backend && yarn fullbuild:webapp && yarn fullbuild:archive",
+    "build-internal": "yarn wsrun --stages --exclude-missing -c build",
+    "build": "dotenv yarn build-internal",
+    "build:local": "dotenv yarn build",
+    "build:webapp-internal": "yarn wsrun -p webapp -c build",
+    "build:webapp": "dotenv yarn build:webapp-internal",
+    "start:dev-internal": "yarn wsrun --parallel --exclude-missing start:dev",
+    "start:dev": "dotenv yarn start:dev-internal",
```
Comment (jakubgs) [inline package.json:21]: This assumes that `dotenv` is available in `PATH`. Why not use `./node_modules/.bin/dotenv`?

File: WebApp/package.json [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+  "scripts": {
+    "analyze:clean": "cross-env rimraf stats.json",
+    "preanalyze": "yarn analyze:clean",
+    "analyze": "node ./internals/scripts/analyze.js",
+    "extract-intl": "node ./internals/scripts/extract-intl.js",
+    "npmcheckversion": "node ./internals/scripts/npmcheckversion.js",
+    "preinstall": "yarn npmcheckversion",
+    "prebuild": "yarn build:clean",
+    "build": "cross-env NODE_ENV=production webpack --config internals/webpack/webpack.prod.babel.js --color -p --progress --hide-modules --display-optimization-bailout",
+    "build:clean": "rimraf ./build",
```
Comment (jakubgs) [inline WebApp/package.json:25]: This also assumes that `rimraf` is avaialble in `PATH`. Use `./node_modules/.bin/rimraf`.

File: Contracts/package.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "dap.ps-contracts",
+  "version": "1.0.0",
+  "description": "Embark contracts",
+  "main": "index.js",
+  "scripts": {
+    "build:testnet": "embark build testnet",
+    "build:development": "embark build development",
+    "build": "embark build livenet",
```
Comment (jakubgs) [inline Contracts/package.json:9]: Same story here. `embark` is not in `PATH`. Use `../node_modules/.bin/embark`.

File: README.md [hunk 4]
```diff
@@ -1,22 +1,107 @@
...
+yarn install
 ```
-or
+
+On Linux, setting up `mongodb` is as easy as `sudo apt install -y mongodb`, which will also start it automatically. You can stop/restart your local DB any time with `sudo systemctl stop mongodb`, or get its status with `sudo systemctl status mongodb`. I recommend using the simple [robo3t](https://robomongo.org/download) to view and edit your DB easily (you'll need to set DApps to `APPROVED` to see `Edit` and `Withdraw` options and to see them appear in Categories etc.)
+
+### Mongo set up
+
+1. `export DB_CONNECTION=mongodb://localhost:27017/mydb`. Make sure you have `DB_CONNECTION` set as an ENV variable so the app knows where to find your local DB.
+2. TODO`yarn run build:localhost`. This will:
```
Comment (jakubgs) [inline README.md:66]: What does the `TODO` mean? Also, it's `yarn run build:local`.

**Review by jakubgs — COMMENTED** [review]
Running `yarn run build:local` fails with
```
 | deploying contracts
 | MiniMeToken already deployed at 0x744d70fdbe2ba4cf95131626614a1763df805b9e
 | Discover already deployed at 0x5bCF2767F86f14eDd82053bfBfd5069F68C2C5F8
 | [DiscoverKyberSwap]: Returned error: Returned error: insufficient funds for gas * price + value
 | Error deploying contracts. Please fix errors to continue.
```

File: README.md (hunk 4 above)
Comment (RyRy79261) [reply README.md:66]: Probably a stray string, will remove

File: Contracts/package.json (hunk 3 above)
Comment (RyRy79261) [reply Contracts/package.json:9]: Noted will amend

File: WebApp/package.json (hunk 2 above)
Comment (RyRy79261) [reply WebApp/package.json:25]: Noted, will amend

File: package.json (hunk 1 above)
Comment (RyRy79261) [reply package.json:21]: When running from within the npm script, its my understanding that it defaults to pulling from the local repository's node modules.

I dont have Embark for example installed globally

**Comment by RyRy79261** [comment]
@jakubgs 

I do agree that potentially a new repo would be preferable as then I can implement a more explicit mono repo.

My biggest hurdle in this rebuild was getting embark to function with the rest of the flow.

Build local for example I couldnt get to function and the contract side was out-of my initial scope

As such my contract addresses are bundled from environment variables instead of the embark config as the contracts that embark would deploy had the wrong constructor parameters leading to reverts.

File: Contracts/package.json (hunk 3 above)
Comment (RyRy79261) [reply Contracts/package.json:9]: As my previous comment, my understand is that when commands are run from within the npm script that it resolves the node modules of the local code base inclusively to the global commands

**Comment by richard-ramos** [comment]
Should `yarn run build:local` deploy to testnet or to mainnet?

Right now, the the package.json for Contracts contains: `"build": "embark build livenet"` meaning that when wsrun command in the root package.json executes, it will attempt to deploy the contract DiscoverKyberSwap to mainnet instead of ropsten (due to the `livenet` parameter of embark,

How can I create a build pointing to testnet/development?

**Comment by richard-ramos** [comment]
I wonder if the Embark issues are product of using v5.2.3. Perhaps they were fixed in 6.0.0 :thinking:

**Comment by RyRy79261** [comment]
@richard-ramos As far as the contracts are concerned, I couldn't get anything beyond compiling to function as expected. 

I created a work around by deploying manually via remix to get around incorrect constructor issues, I then pull the addresses from environment variables at build

Building for prod: `yarn create:fullbuild`
Running local: `yarn start:dev`

**Comment by richard-ramos** [comment]
This PR should fix the issue with Embark deployment: https://github.com/RyRy79261/discover/pull/1
I was able to execute `embark run testnet` and have the contracts deployed (after removing them from the testnet {} section of contracts.js

**Comment by RyRy79261** [comment]
Thanks @richard-ramos I see that `library: 'embarkjs',` is specified as the library, is Ethers an option? I used it in lieu of Embark for the rebuild

**Comment by jakubgs** [comment]
I'm still failing to build the app with `yarn build:local` due to:
```
 | [DiscoverKyberSwap]: Returned error: Returned error: insufficient funds for gas * price + value
 | Error deploying contracts. Please fix errors to continue.
```
And using `yarn start:dev` fails on page load with;
```
 | Error: ENOENT: no such file or directory, stat '/home/sochan/work/dapps:Backend/discover/Backend/frontend/index.html'
```

**Comment by RyRy79261** [comment]
@jakubgs Seems theres a pathing issue on your machine, build local is deprecated as I am unsure what its initial purpose was in the repo since it dealt with the embark side of things

**Comment by jakubgs** [comment]
@RyRy79261 yeah, we'll need to clean up the legacy targets from `package.json` and `Makefile` after this is moved to a separate repo.
